packages/generator-langium/templates/test/test/linking/linking.test.ts

```diff
@@ -0,0 +1,53 @@
+import { afterEach, beforeAll, describe, expect, test } from "vitest";
+import { EmptyFileSystem, type LangiumDocument } from "langium";
+import { expandToString as s} from "langium/generate";
+import { clearDocuments, parseHelper } from "langium/test";
+import { create<%= LanguageName %>Services } from "../../src/language/<%= language-id %>-module.js";
+import { Model, isModel } from "../../src/language/generated/ast.js";
+
+let services: ReturnType<typeof create<%= LanguageName %>Services>;
+let parse:    ReturnType<typeof parseHelper<Model>>;
+let document: LangiumDocument<Model> | undefined;
+
+beforeAll(async () => {
+    services = create<%= LanguageName %>Services(EmptyFileSystem);
+    parse = parseHelper<Model>(services.<%= LanguageName %>);
+
+    // activate the following if your linking test requires elements from a built-in library, for example
+    // await services.shared.workspace.WorkspaceManager.initializeWorkspace([]);
+});
+
+afterEach(async () => {
+    document && clearDocuments(services, [ document ]);
+});
+
+describe('Linking tests', () => {
+
+    test('linking of greetings', async () => {
+        document = await parse(`
+            person Langium
+            Hello Langium!
+        `);
+
+        expect(
+            // here we first check for validity of the parsed document object by means of the reusable function
+            //  'checkDocumentValid()' to sort out (critical) typos first,
+            // and then evaluate the cross references we're interested in by checking
+            //  the referenced AST element as well as for a potential error message;
+            checkDocumentValid(document)
+                || document.parseResult.value.greetings.map(g => g.person.ref?.name || g.person.error?.message).join('\n')
```

Comment:
This line is too generic, it covers all cases where a person ref is not resolved correctly. I would recommend to address the exact element in the model with multiple expect().

```ts
expectNoErrors(document); //see below
expect(document.parseResult.value.greetings).toHaveLength(1);
expect(document.parseResult.value.greetings[0].person.ref?.name, document.parseResult.value.greetings[0].person.error?.message).toBeDefined();
```

Reply:
First of all: Don't forget it's just an example!

> This line is too generic

Why? The number of expected Greetings is (implicitly) specified in the expected string representation.
It just doesn't follow the "established pattern".

> `expect(document.parseResult.value.greetings).toHaveLength(1);`

Actually I don't like such kinds of tests: Why? It gives you failure messages like this:
```
AssertionError: expected [ { '$type': 'Greeting', …(5) }, …(1) ] to have a length of 1 but got 2
```
IMO, this is not developer friendly, esp. if the test pops up half a year after writing for somebody else than the original author. Imagine a scenario that is not as trivial as this one.

Another aspect is that the test stops immediately at the cardinality failure, nothing else is executed.
You have to investigate and fix it (or deactivate) to come to the next stumble stone.

My proposed solution is more tolerant on that, but - agreed - you have take a few seconds and think different.

![image](https://github.com/eclipse-langium/langium/assets/8513883/72efdec9-2fc7-4fbc-bcd7-ebc7fc4a3d16)

Give it try 😉

Reply:
I see, look at my other comment. I will resolve this thread.